libssh2/src/Network/SSH/Client/LibSSH2.hs

```diff
@@ -216,3 +226,124 @@ withChannelBy createChannel extractChannel actions = do
   freeChannel ch
   return (exitStatus, result)
 
+-- | Execute some actions within SFTP connection.
+-- Uses public key authentication.
+withSFTP :: FilePath          -- ^ Path to known_hosts file
+         -> FilePath          -- ^ Path to public key file
+         -> FilePath          -- ^ Path to private key file
+         -> String            -- ^ Passphrase
+         -> String            -- ^ Remote user name
+         -> String            -- ^ Remote host name
+         -> Int               -- ^ Remote port number (usually 22)
+         -> (Sftp -> IO a)    -- ^ Actions to perform on sftp session
+         -> IO a
+withSFTP known_hosts public private passphrase login hostname port fn =
+  withSession hostname port $ \s -> do
+    r <- checkHost s hostname port known_hosts
+    when (r == MISMATCH) $
+      error $ "Host key mismatch for host " ++ hostname
+    publicKeyAuthFile s login public private passphrase
+    withSftpSession s fn
+
+-- | Execute some actions within SFTP connection.
+-- Uses username/password authentication.
+withSFTPUser :: FilePath          -- ^ Path to known_hosts file
+             -> String            -- ^ Remote user name
+             -> String            -- ^ Remote password
+             -> String            -- ^ Remote host name
+             -> Int               -- ^ Remote port number (usually 22)
+             -> (Sftp -> IO a)    -- ^ Actions to perform on sftp session
+             -> IO a
+withSFTPUser known_hosts login password hostname port fn =
+  withSession hostname port $ \s -> do
+    r <- checkHost s hostname port known_hosts
+    when (r == MISMATCH) $
+      error $ "Host key mismatch for host " ++ hostname
+    usernamePasswordAuth s login password
+    withSftpSession s fn
+
+-- | Execute some actions within SFTP session
+withSftpSession :: Session           -- ^ Remote host name
+                -> (Sftp -> IO a)    -- ^ Actions to perform on sftp session
+                -> IO a
+withSftpSession session =
+  E.bracket (sftpInit session) sftpShutdown
+
+type SftpList = [(BSS.ByteString, SftpAttributes)]
+
+-- | Reads directory information
+-- Returns the list of files with attributes, directory . and ..
+-- are not excluded
+sftpListDir :: Sftp        -- ^ Opened sftp session
+            -> FilePath    -- ^ Remote directory to read
+            -> IO SftpList
+sftpListDir sftp path =
+  withDirList sftp path $ \h ->
+    collectFiles h []
+
+withDirList :: Sftp
+            -> FilePath
+            -> (SftpHandle -> IO a)
+            -> IO a
+withDirList sftp path = E.bracket (sftpOpenDir sftp path) sftpCloseHandle
+
+collectFiles :: SftpHandle -> SftpList -> IO SftpList
```

Comment:
Is this function to be used anywhere else? It looks like almost trivial helper, so maybe it is better to merge it into sftpListDir under "where"?

Reply:
Sure, done.